Talos2 2nd cpu (#491)
3mdeb/news-and-ideas: blog/content/post/2022-04-12-talos2_2nd_cpu_and_testing.md

---
title: Talos II - second CPU support and test automation
abstract: 'Another post about our adventures with porting coreboot for Talos II.
          This phase focused on enabling second CPU and its internal devices.
          We also expanded our test suite.'
cover: /covers/coreboot-logo.svg
author: krystian.hebel
layout: post
published: true
date: 2022-04-12
archives: "2022"

tags:
  - coreboot
  - firmware
  - openpower
categories:
  - Firmware

---

Another post about our adventures with porting coreboot for Talos II. This phase
focused on enabling second CPU and its internal devices. We also expanded our
test suite.

## Links to binaries and installation instructions

> TBD

Reviewer: @krystian-hebel We already know the links. Maybe we can just link to the relevent section in the documentation?

Author: Added

switched away from FIT payload and started using ELF. Main motivation for this
change was fact that [FIT decompression code does not work as expected](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/6EZWU7YPUJE564GNCV7U32IXWPFTV7FB/)
in coreboot. With payload compressed with LZMA, flash memory footprint of whole
coreboot and Skiboot has shrunk to just below 700 KB. For comparison, Hostboot
takes about 30 MB of space.

## New tests

In order to achive greater reliability of the firmware and to avoid bugs test
base on the Talos II platform has been improved.

Currently, before each firmware release, the following test suites will be
performed:
- [Dasharo compatibility: coreboot base port][CBP], which contains test cases
  for checking correctness of coreboot porting on the device.
- [Dasharo compatibility: Petitboot payload support][PBT], which contains test
  cases for checking Petitboot availabity and functionality.
- [Dasharo compatibility: Heads bootloader support][HDS], which contains test
  cases for checking Heads bootloader availabity.
- [Dasharo compatibility: USB detection][USB], which contains test cases for

>   checking correctness of USB detection after coldboot, warmoot and system

Reviewer 2: ```suggestion
  checking correctness of USB detection after coldboot, warmboot and system
```

  reboot.
- [Dasharo compatibility: USB booting][UBB], which contains tests cases for
  checking correctness of booting from USB after coldboot, warmboot and system
  reboot.
- [Dasharo compatibility: Debian Stable and Ubuntu LTS support][LBT], which
  contains test cases for checking correctness of installing and booting into
  installed OS.
- [Dasharo compatibility: Device Tree][DVT], which contains test cases for

>   checking correctness of presentation coreboot in Device Tree.

Reviewer 2: ```suggestion
  checking correctness of coreboot presentation in Device Tree.
```

  checking correctness of CPU work.

All test cases documentation is available under this [link][Tests].
Full test matrix for Talos II is available under this [link][Matrix]

[CBP]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/100-coreboot-base-port/
[PBT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31V-petitboot-payload-support/
[HDS]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31U-heads-bootloader-support/
[USB]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31O-usb-detect/
[UBB]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31N-usb-boot/
[LBT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/308-debian-stable-and-ubuntu-lts-support/
[DVT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31W-device-tree/
[CPU]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31T-cpu-status/
[Tests]: https://docs.dasharo.com/unified-test-documentation/overview/
[Matrix]: https://docs.dasharo.com/variants/talos_2/test-matrix/

## Summary

> TBD

Reviewer: @krystian-hebel What do we want to include here?

Author: Added

  takes precedence in setting the other. Complex calculations are done during
  the initialization to choose "more important" clock source - the one with
  higher delay, as it is more inert.
- memory controller - even though memory initialization and training is repeated
  for secondary processor, only main CPU sets a common memory space address
  translation maps.
- PCIe controller - links are split differently between ports connected to each
  of processor sockets.

In addition to various initialization steps, we had to update device tree passed
to Skiboot to include information about new CPU and devices it adds. We also
switched away from FIT payload and started using ELF. Main motivation for this
change was fact that [FIT decompression code does not work as expected](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/6EZWU7YPUJE564GNCV7U32IXWPFTV7FB/)
in coreboot. With payload compressed with LZMA, flash memory footprint of whole
coreboot and Skiboot has shrunk to just below 700 KB. For comparison, Hostboot
takes about 30 MB of space.

## New tests

In order to achive greater reliability of the firmware and to avoid bugs test

> base on the TalosII platform has been improved.

Author: ```suggestion
base on the Talos II platform has been improved.
```

  checking correctness of CPU work.

All test cases documentation is available under this [link][Tests].
Full test matrix for Talos II is available under this [link][Matrix]

[CBP]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/100-coreboot-base-port/
[PBT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31V-petitboot-payload-support/
[HDS]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31U-heads-bootloader-support/
[USB]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31O-usb-detect/
[UBB]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31N-usb-boot/
[LBT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/308-debian-stable-and-ubuntu-lts-support/
[DVT]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31W-device-tree/
[CPU]: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31T-cpu-status/
[Tests]: https://docs.dasharo.com/unified-test-documentation/overview/
[Matrix]: https://docs.dasharo.com/variants/talos_2/test-matrix/

> Description of all dasharo test cases

Author: ???

Reviewer 3: Redundant line. I just deleted it

---
title: Talos II - second CPU support and test automation
abstract: 'Another post about our adventures with porting coreboot for Talos II.
          This phase focused on enabling second CPU and its internal devices.
          We also expanded our test suite.'
cover: /covers/coreboot-logo.svg
author: krystian.hebel
layout: post
published: true

> date: 2022-04-12

Reviewer: ```suggestion
date: 2022-04-15
```

- TOD (time of day) - in case of discrepancy between two clocks one of them
  takes precedence in setting the other. Complex calculations are done during
  the initialization to choose "more important" clock source - the one with
  higher delay, as it is more inert.
- memory controller - even though memory initialization and training is repeated
  for secondary processor, only main CPU sets a common memory space address
  translation maps.
- PCIe controller - links are split differently between ports connected to each
  of processor sockets.

In addition to various initialization steps, we had to update device tree passed
to Skiboot to include information about new CPU and devices it adds. We also
switched away from FIT payload and started using ELF. Main motivation for this
change was fact that [FIT decompression code does not work as expected](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/6EZWU7YPUJE564GNCV7U32IXWPFTV7FB/)
in coreboot. With payload compressed with LZMA, flash memory footprint of whole
coreboot and Skiboot has shrunk to just below 700 KB. For comparison, Hostboot
takes about 30 MB of space.

## New tests

> TBD

Reviewer: @sulewskiprzemyslaw We would expect short description of changes in the test area here.
